packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public view returns (bool) {
+    require(!revokedSigners[signer], "Signer has been revoked");
+    require(
+      getAccounts().attestationSignerToAccount(signer) == issuer,
+      "Signer has not been authorized as an AttestationSigner by the issuer"
+    );
+    bytes32 structHash = keccak256(
+      abi.encode(EIP712_VALIDATE_ATTESTATION_TYPEHASH, identifier, issuer, account, issuedOn)
+    );
+    address guessedSigner = Signatures.getSignerOfTypedDataHash(
+      eip712DomainSeparator,
+      structHash,
+      v,
+      r,
+      s
+    );
+    return guessedSigner == signer;
+  }
+
+  /**
+   * @notice Registers an attestation with a valid signature
+   * @param identifier Hash of the identifier to be attested
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account being mapped to the identifier
+   * @param issuedOn Time at which the issuer issued the attestation in Unix time 
+   * @param signer Address of the signer of the attestation
+   * @param v The recovery id of the incoming ECDSA signature
+   * @param r Output value r of the ECDSA signature
+   * @param s Output value s of the ECDSA signature
+   * @dev Throws if sender is not the issuer, account, or an authorized AttestationSigner
+   * @dev Throws if an attestation with the same (identifier, issuer, account) already exists
+   */
+  function registerAttestation(
+    bytes32 identifier,
+    address issuer,
+    address account,
+    uint256 issuedOn,
+    address signer,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public isValidUser(issuer, account) {
+    require(
+      isValidAttestation(identifier, issuer, account, issuedOn, signer, v, r, s),
+      "Signature is invalid"
+    );
+    for (uint256 i = 0; i < identifierToAddresses[identifier][issuer].length; i = i.add(1)) {
+      // This enforces only one attestation to be uploaded
+      // for a given set of (identifier, issuer, account)
+      // Editing/upgrading an attestation requires that it be deleted before a new one is registered
+      require(
+        identifierToAddresses[identifier][issuer][i].account != account,
+        "Attestation for this account already exists"
+      );
+    }
+    OwnershipAttestation memory attestation = OwnershipAttestation(account, issuedOn, signer);
+    identifierToAddresses[identifier][issuer].push(attestation);
+    addressToIdentifiers[account][issuer].push(identifier);
+    emit AttestationRegistered(identifier, issuer, account, issuedOn, signer);
+  }
+
+  /**
+   * @notice Deletes an attestation 
+   * @param identifier Hash of the identifier to be deleted
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account mapped to the identifier
+   * @dev Throws if sender is not the issuer, account, or an authorized AttestationSigner
+   */
+  function deleteAttestation(bytes32 identifier, address issuer, address account)
```

Comment:
Note that the same attestation may be re-uploaded to add this attestation back to the chain. If an issuer, for example, is trying to revoke an attestation, using this function will not work because the user can always re-add it. If anyone is allowed to upload an attestation, which I generally suggest, then anyone can reupload an attestation they extract from transaction logs. I would suggest we add a `mapping (bytes32 => bool) revokedAttestations` that we set when deleting an attestation. This array should be checked before an upload is accepted. The mapping key needs to be unique to an attestation, and permit an issuer to reissue an attestation. Using the `structHash` as in the signature check works nicely for this since it includes the issuer, account, and a timestamp. If the issuer wants to reissue the attestation they can, and it will have a new timestamp.